clients/core/src/main/scala/mqperf/Sender.scala

```diff
@@ -19,49 +21,107 @@ class Sender(config: Config, mq: Mq, clock: Clock) extends StrictLogging {
-          val msgsToSend = math.min(config.msgsPerSecond, permits.getAndUpdate(p => math.max(0, p - config.msgsPerSecond)))
-          logger.info(s"Messages to send: $msgsToSend")
-          val batches = msgsToSend / batchSize
-          (1 to batches).foreach { _ =>
-            val sendStart = clock.millis()
-            mqSender.send(nextMessagesBatch()).onComplete { result =>
-              permits.addAndGet(batchSize)
-              LocalMetrics.messageCounter.inc(batchSize.toDouble)
-              LocalMetrics.messageLatencyHistogram.observe((clock.millis() - sendStart).toDouble)
-              result match {
-                case Failure(t) => logger.error("Exception when sending a batch of messages", t)
-                case Success(_) =>
-              }
+    val senderRun = runParallel(mqSender, config.senderConcurrency) >>
+      IO(logger.info("Sending done, waiting for all messages to be flushed ...")) >>
+      IO.sleep(3.seconds) >>
+      IO.fromFuture(IO(mqSender.close()))
+
+    senderRun
+      .handleError(e => logger.error("Sender failed", e))
+      .unsafeToFuture()
+  }
+
+  private def runParallel(mqSender: MqSender, senderConcurrency: Int): IO[Unit] = {
+    val batchSize = config.batchSizeSend
+
+    def sendRunner(state: Ref[IO, State], iterationNo: Int = 0): IO[Unit] = {
+      def iteration(state: Ref[IO, State]): IO[Unit] = {
+        state.get
+          .flatMap(s =>
+            if (s.sent >= s.permits) {
+              IO.unit
+            } else {
+              for {
+                sendStart <- IO(clock.millis())
+                _ <- IO.fromFuture(IO(mqSender.send(nextMessagesBatch(batchSize))))
+                _ <- state.update { s => s.copy(sent = s.sent + batchSize) }
+                _ <- IO {
+                  LocalMetrics.messageCounter.inc(batchSize.toDouble)
+                  LocalMetrics.messageLatencyHistogram.observe((clock.millis() - sendStart).toDouble)
+                }
+                result <- iteration(state)
+              } yield result
             }
+          )
+          .handleErrorWith { e =>
+            IO(logger.error(s"Sender iteration[$iterationNo] failed ", e)) >>
+              iteration(state)
           }
+      }
 
-          val iterationEnd = clock.millis()
-          // a whole iteration should take at least a second
-          Thread.sleep(math.max(0, 1000L - (iterationEnd - iterationStart)))
+      state.get.flatMap(s => {
+        if (iterationNo < s.currentIteration) {
+          iteration(state) >> sendRunner(state, iterationNo + 1)
+        } else {
+          s.nextIterationHook.get.flatMap(isTestEnd => {
+            if (isTestEnd) IO.unit
+            else iteration(state) >> sendRunner(state, iterationNo + 1)
+          })
         }
+      })
+    }
 
-        logger.info("Sending done, waiting for all messages to be flushed ...")
-        Thread.sleep(3000L)
+    def loop(until: Long, state: Ref[IO, State], iterationNo: Int = 0): IO[Unit] = {
```

Comment:
there's a lot of state-sharing and understanding how this works w/deferred isn't that easy ;) I'm still not sure if it's correct

maybe it would be simpler if we had a bunch of sender processes, looping with acquiring permits from a semaphore and sending a batch of messages, plus a controller process which replenishes the semaphore up to `msgsPerSecond` every second? 

after the test finishes, we could simply interrupt all the sender fibers

Reply:
one place where I'm not sure if the current design works correctly is what happens if some fiber lags behind for one iteration and its local `iterationNo` gets out of sync. There will be some excess `iteration` calls I think